ipykernel/parentpoller.py

```diff
@@ -22,19 +22,41 @@ class ParentPollerUnix(Thread):
     when the parent process no longer exists.
     """
 
-    def __init__(self):
-        """Initialize the poller."""
+    def __init__(self, parent_pid=0):
+        """Initialize the poller.
+
+        Parameters
+        ----------
+        parent_handle : int, optional
+            If provided, the program will terminate immediately when
+            process parent is no longer this original parent.
+        """
         super().__init__()
+        self.parent_pid = parent_pid
         self.daemon = True
 
     def run(self):
         """Run the poller."""
         # We cannot use os.waitpid because it works only for child processes.
         from errno import EINTR
 
+        # before start, check if the passed-in parent pid is valid
```

Comment:
Maybe this whole thing is overly complicated? We could also elect to just trust JPY_PARENT_PID directly - the only problem is if it's a stale environment variable from somewhere.

Reply:
Thanks for merging!

My updated thoughts on this comment here is that it's good that we check if JPY_PARENT_PID is the actual parent. It's actually possible (is it common?) to have situations like  `papermill -> bash -> ipykernel` or with similar intermediate processes, where JPY_PARENT_PID will point to papermill but the ipykernel ppid will be bash.